libs/json-rpc-stdio/src/lib.rs

```diff
@@ -1,35 +1,133 @@
-use jsonrpc_core::IoHandler;
-use tokio::io::{AsyncBufReadExt, AsyncRead, AsyncWrite, AsyncWriteExt};
+use jsonrpc_core::{IoHandler, MethodCall, Request};
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::HashMap,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc,
+    },
+};
+use tokio::{
+    io::{AsyncBufReadExt, AsyncRead, AsyncWrite, AsyncWriteExt},
+    sync::{mpsc, oneshot},
+};
 
-pub async fn run(handler: &IoHandler) -> std::io::Result<()> {
-    run_with_io(handler, tokio::io::stdin(), tokio::io::stdout()).await
+#[derive(Debug)]
+pub struct Client(Arc<ClientInner>, AtomicU64);
+
+impl Client {
+    pub fn new() -> (Client, ClientAdapter) {
+        let (request_sender, request_receiver) = mpsc::channel(30);
+        let client = Client(Arc::new(ClientInner { request_sender }), Default::default());
+        let adapter = ClientAdapter { request_receiver };
+
+        (client, adapter)
+    }
+
+    pub async fn call<Req, Res>(&self, method: String, params: Req) -> jsonrpc_core::Result<Res>
+    where
+        Req: serde::Serialize,
+        Res: serde::de::DeserializeOwned,
+    {
+        let id = self.1.fetch_add(1, Ordering::Relaxed);
```

Comment:
This is personal style but I don't really like the `self.1`. I would rather we named the fields in the struct. It makes it a lot clearer what we acting on.

Reply:
Yep that's just lazy on my part — will do a named struct.